### DISCUSSION TITLE
Help Request: Model Training with Neural Hydrology Framework - Loss Instability + Poor Test Set Performance

### DISCUSSION THREAD

**Discussion by Zhangxicicici** [neuralhydrology/neuralhydrology#252] [Q&A]
> Hello everyone! I'm encountering critical issues while training a neural hydrology model for streamflow prediction using the Neural Hydrology framework. The model fails to converge, and the test set predictions are extremely poor. I'd greatly appreciate your expertise in resolving these problems.
> ### **Problem Description**
> **1. Training Phase Anomalies**
> During training, the model's loss does not decrease stably. Observations from TensorBoard's TIMESERIES panel:
> 
> Training Set:
> Metrics like train/avg_loss and train/avg_total_loss fluctuate wildly without a downward trend.
> train/loss and train/total_loss show significant oscillations, indicating the model struggles to learn.
> Validation Set:
> valid/avg_loss and valid/avg_total_loss initially decline but plateau early, raising doubts about true convergence.
> valid/mean_nse and valid/median_nse improve marginally, but the gains are inconsistent with training instability.
> **2. Test Phase Performance**
> The test set results for streamflow prediction are catastrophic, with an NSE (Nash-Sutcliffe Efficiency) of -1.562. Key issues:
> 
> Temporal Coverage Discrepancy:
> The predicted curve (orange) only appears between 2007-2008, despite the test set spanning 2006-2007 (as configured).
> Poor Fit:
> The predicted curve fails to capture critical flow characteristics (e.g., peak flows, seasonal patterns), deviating significantly from the observed values (blue).
> Magnitude Mismatch:
> In 2006, predicted values are near zero or noise-level, while observed flows exhibit substantial variability (peaks >50 mm/d).
> 
> ### Supporting Information
> **1. Key Visualizations**
> TensorBoard Training/Validation Curves:
> 
<img width="647" alt="{1EB62BD3-7EAB-4149-86E8-F9098296C34E}" src="https://github.com/user-attachments/assets/cf7c03fb-5b52-4f00-a4d1-5c6d0b263f5c" />

> Test Set Prediction Results:
> ![image](https://github.com/user-attachments/assets/cac69aa6-8e71-44c1-8c2c-74eb79a23728)
```
NSE: -0.072
MSE: 6.393
RMSE: 2.528
KGE: -0.381
Alpha-NSE: 0.148
Beta-KGE: 0.823
Beta-NSE: -0.171
Pearson-r: -0.073
FHV: -77.932
FMS: -81.144
FLV: 64.660
Peak-Timing: 2.200
Peak-MAPE: 82.484
```
> X-axis: Time (2006-2008)
> Y-axis: Discharge (mm/d)
> Blue: Observed streamflow
> Orange: Model predictions
> Critical Issue: Predictions vanish in 2006 despite valid time indices.
> **2. Temporal Configuration Analysis**

```
run_dir: /home/users/zhangxi/knowledge-graph/run

train_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt
validation_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt
test_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt

train_start_date: "05/01/2007"
train_end_date: "31/12/2010"
validation_start_date: "10/01/2010"
validation_end_date: "1/12/2011"
test_start_date: "01/01/2006"
test_end_date: "01/12/2007"

device: cuda:1

validate_every: 3

validate_n_random_basins: 1

metrics:
  - NSE

model: cudalstm

head: regression

output_activation: linear

hidden_size: 20

initial_forget_bias: 3

output_dropout: 1

optimizer: Adam

loss: MSE

learning_rate:
  0: 1e-3
  5: 5e-05
  15: 5e-5
  30: 1e-5
  50: 1e-5

batch_size: 32

epochs: 50

clip_gradient_norm: 0.5

predict_last_n: 1

seq_length: 365

num_workers: 8

log_interval: 5

log_tensorboard: True

log_n_figures: 1

save_weights_every: 1

dataset: camels_us

data_dir: /home/users/zhangxi/knowledge-graph/basin_timeseries_v1p2_metForcing_obsFlow/CAMELS_US

forcings:
  - daymet
dynamic_inputs:
  - prcp(mm/day)
  - tmax(C)
  - tmin(C)

target_variables:
  - QObs(mm/d)

clip_targets_to_zero:
  - QObs(mm/d)
```

**Comment by gauchm** [comment]
I see that your config sets output_dropout to 1. You might want to read about [Dropout](https://towardsdatascience.com/dropout-in-neural-networks-47a162d621d9/) to understand the implications of this. Essentially, during training, your are removing all information before asking the model to make a prediction. So it's clear that the model can't learn anything from that.

**Reply by Zhangxicicici** [reply]
First off, huge thanks for the previous replies! After making adjustments based on the suggestions, I did see notable improvements in the training set performance. However, the results are still far from satisfactory. The training set has improved significantly, but the second issue I mentioned earlier - the poor convergence of the validation and test sets - remains a big problem. No matter how many training epochs I run, the NSE for these sets stays below 0.5. I'm really desperate for more advice on this. Here's my newly modified configuration code, and I'm extremely eager for your insights.
It's important to note that the watershed data I'm using is not from the CAMELS-US dataset; it's just placed in the corresponding directory.
**1. Current Situation Summary**
Training Set: Shows significant improvement, indicating the model has the ability to learn patterns from the data.
Validation & Test Sets: NSE consistently below 0.5. 
<img width="960" alt="{FB184332-536C-4616-A227-76CA43E9360C}" src="https://github.com/user-attachments/assets/11e4ae89-d45a-4538-bc7f-b88763169937" />

**2. Newly Modified Configuration Code (Relevant Snippets)**
```
# --- Experiment configurations --------------------------------------------------------------------

# experiment name, used as folder name
experiment_name: test_run
run_dir: /home/users/zhangxi/knowledge-graph/run
# files to specify training, validation and test basins (relative to code root or absolute path)
train_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt
validation_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt
test_basin_file: /home/users/zhangxi/knowledge-graph/neuralhydrology-master/examples/01-Introduction/1_basin.txt

# if you want to use different (continuous or split) periods per basin (and period) define path to pickle files here.

# Use different (continuous or split) periods for each basin (defined by pickle files)
#per_basin_train_periods_file: /home/users/zhangxi/66666666_train_periods.pkl
#per_basin_validation_periods_file: /home/users/zhangxi/6666666validation_data.pkl
#per_basin_test_periods_file: /home/users/zhangxi/66666666test_data.pkl


# training, validation and test time periods (format = 'dd/mm/yyyy')
train_start_date: "05/01/2006"
train_end_date: "31/12/2008"
validation_start_date: "10/01/2010"
validation_end_date: "1/12/2011"
test_start_date: "01/01/2009"
test_end_date: "01/12/2009"

# which GPU (id) to use [in format of cuda:0, cuda:1 etc, or cpu, mps or None]
device: cuda:1
 
# --- Validation configuration ---------------------------------------------------------------------

# specify after how many epochs to perform validation
validate_every: 3


# specify how many random basins to use for validation
validate_n_random_basins: 1

# specify which metrics to calculate during validation (see neuralhydrology.evaluation.metrics)
# this can either be a list or a dictionary. If a dictionary is used, the inner keys must match the name of the
# target_variable specified below. Using dicts allows for different metrics per target variable.
metrics:
  - NSE

# --- Model configuration --------------------------------------------------------------------------

# base model type [lstm, ealstm, cudalstm, embcudalstm, mtslstm]
# (has to match the if statement in modelzoo/__init__.py)
model: cudalstm

# prediction head [regression]. Define the head specific parameters below
head: regression

# ----> Regression settings <----
output_activation: linear


# ----> General settings <----

# Number of cell states of the LSTM
hidden_size: 20

# Initial bias value of the forget gate
initial_forget_bias: 3

# Dropout applied to the output of the LSTM
output_dropout: 0.5

# --- Training configuration -----------------------------------------------------------------------

# specify optimizer [Adam]
optimizer: Adam

# specify loss [MSE, NSE, RMSE]
loss: MSE

# specify learning rates to use starting at specific epochs (0 is the initial learning rate)
learning_rate:
  0: 1e-3
  50: 5e-05
  80: 5e-5
  #30: 1e-5
  #50: 1e-5
  120: 5e-6
  160: 3e-6
  200: 1e-6
  250: 5e-7
  300: 3e-7
  350: 1e-7


# Mini-batch size
batch_size: 32

# Number of training epochs
epochs: 1000

# If a value, clips the gradients during training to that norm.
clip_gradient_norm: 0.5

# Defines which time steps are used to calculate the loss. Can't be larger than seq_length.
# If use_frequencies is used, this needs to be a dict mapping each frequency to a predict_last_n-value, else an int.
predict_last_n: 1

# Length of the input sequence
# If use_frequencies is used, this needs to be a dict mapping each frequency to a seq_length, else an int.
seq_length: 365

# Number of parallel workers used in the data pipeline
num_workers: 8

# Log the training loss every n steps
log_interval: 5

# If true, writes logging results into tensorboard file
log_tensorboard: True

# If a value and greater than 0, logs n random basins as figures during validation
log_n_figures: 1

# Save model weights every n epochs
save_weights_every: 1

# --- Data configurations --------------------------------------------------------------------------

# which data set to use [camels_us, camels_gb, global, hourly_camels_us]
dataset: camels_us

# Path to data set root
data_dir: /home/users/zhangxi/knowledge-graph/basin_timeseries_v1p2_metForcing_obsFlow/CAMELS_US

# Forcing product [daymet, maurer, maurer_extended, nldas, nldas_extended, nldas_hourly]
# can be either a list of forcings or a single forcing product
forcings:
 # - maurer
  - daymet
 # - nldas

dynamic_inputs:
 # - PRCP(mm/day)_nldas
 # - PRCP(mm/day)_maurer
  - prcp(mm/day)
  #- srad(W/m2)_daymet
  - tmax(C)
  - tmin(C)
  #- vp(Pa)_daymet
  #- swe(mm)

# which columns to use as target
target_variables:
  - QObs(mm/d)

# clip negative predictions to zero for all variables listed below. Should be a list, even for single variables.
clip_targets_to_zero:
  - QObs(mm/d)
```


I'm really counting on your experiences and expertise. Any suggestions, no matter how small, could be a huge help in getting this model to perform better on validation and test sets. Thanks so much in advance!

**Reply by gauchm** [reply]
It's hard to give specific advice without knowing the setting (how many train/test basins, how many samples per basin, ...), but I suggest you play around a bit with the hyperparameters. 

Specifically, it looks like the val performance is peaking already at the very beginning of training. So maybe set validate_every to 1, play around with (lower) learning rates, maybe also different hidden_size and dropout (0.5 is still pretty high, although that is probably not the cause for noisy validation results).

Also, maybe more importantly: is this a single-basin setting? I.e., are you training and evaluating on a single basin? If it isn't, and you are training and evaluating on multiple basins, then you should set validate_n_random_basins to a (much) larger value. Currently, each validation run on just one random basin (a different one each time), so it's no wonder the performance is very noisy.

**Reply by Zhangxicicici** [reply]
Thank you very much for your responses! I'd like to further clarify the input features of my model: the current input variables include daily precipitation, daily maximum temperature, and daily minimum temperature—all daily-scale observational data from this single basin (spanning six years).
Thank you again for your help!
